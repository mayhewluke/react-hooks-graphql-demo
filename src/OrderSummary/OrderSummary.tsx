import gql from "graphql-tag";
import React, { useState } from "react";
import { Query } from "react-apollo";

import AddPizza from "../AddPizza/AddPizza";
import LineItem from "../LineItem/LineItem";

export const SIZES_QUERY = gql`
  {
    pizzaSizes {
      name
      basePrice
    }
  }
`;

interface LineItem {
  size: string;
  price: number;
  toppings: string[];
}

interface SizesQueryData {
  pizzaSizes: Array<{ name: string; basePrice: number }>;
}

const renderOrder = (pizzas: LineItem[]) => (
  <table id="order-summary">
    <thead>
      <tr>
        <th>Size</th>
        <th>Price</th>
      </tr>
    </thead>
    <tbody>
      {pizzas.map(({ size, price }, index) => (
        <LineItem key={index} size={size} cost={price} />
      ))}
    </tbody>
  </table>
);

const OrderSummary = () => {
  const [pizzas, setPizzas] = useState([] as LineItem[]);
  return (
    <Query<SizesQueryData> query={SIZES_QUERY}>
      {({ loading, error, data }) => (
        <>
          <div data-testid="order">
            {pizzas.length ? renderOrder(pizzas) : <p>No pizzas ordered yet</p>}
          </div>
          {loading && <p>Loading...</p>}
          {error && <p>{error.toString()}</p>}
          {data && data.pizzaSizes && (
            <AddPizza
              pizzaSizes={data.pizzaSizes}
              addPizza={pizza => setPizzas([...pizzas, pizza])}
            />
          )}
        </>
      )}
    </Query>
  );
};

export default OrderSummary;
