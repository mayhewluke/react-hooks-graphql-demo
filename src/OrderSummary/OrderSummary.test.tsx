import "jest-dom/extend-expect";
import "react-testing-library/cleanup-after-each";

import React from "react";
import { MockedProvider, MockedResponse } from "react-apollo/test-utils";
import { render as rtlRender, wait } from "react-testing-library";

import OrderSummary, { SIZES_QUERY } from "./OrderSummary";

jest.mock("../LineItem/LineItem", () => jest.fn(() => false));

describe("OrderSummary", () => {
  const render = (response: Partial<MockedResponse>) =>
    rtlRender(
      <MockedProvider
        mocks={[
          {
            request: {
              query: SIZES_QUERY,
            },
            ...response,
          },
        ]}
        addTypename={false}
      >
        <OrderSummary />
      </MockedProvider>,
    );
  const pizzaSizes = [
    { name: "mini", basePrice: 1 },
    { name: "jumbo", basePrice: 2 },
  ];

  it("shows a loading message while pizza info is loading", () => {
    const { container } = render({ result: { data: { pizzaSizes } } });
    expect(container).toHaveTextContent(/loading/i);
  });

  it("shows an error message if the network fails", async () => {
    const { container } = render({ error: new Error("Something went wrong!") });
    await wait(() => expect(container).toHaveTextContent(/error/i));
  });

  it("shows an error message if the query fails", async () => {
    const { container } = render({
      result: { errors: [{ message: "Error!" }] },
    });
    await wait(() => expect(container).toHaveTextContent(/error/i));
  });

  it("displays a message when no pizzas have been added yet", async () => {
    const { getByTestId } = render({
      result: { data: { pizzaSizes } },
    });
    await wait(() =>
      expect(getByTestId("order").textContent).toMatchSnapshot(),
    );
  });
});
