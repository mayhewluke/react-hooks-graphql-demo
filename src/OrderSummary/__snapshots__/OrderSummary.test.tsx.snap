// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderSummary displays a message when no pizzas have been added yet 1`] = `"No pizzas ordered yet"`;
